!!! FEATURE: References on creation and Copy

kitsunet opened the pull request:
This reworks references so that multiple reference properties can be set via a single command and also references can be attached to `CreateNodeAggregateWithNode` which is also used for copying nodes.

The serialization format is adapted to allow multiple reference properties, which also affects all behat tests with references.

To migrate existing events run:

`./flow migrateevents:migratesetreferencestomultinameformat`

Marked breaking as it requires an event migration.

This requires https://github.com/neos/neos-ui/pull/3844 for the UI

A behat test to show reference copies work was added.

**Upgrade Instructions**

The API to set references changed a little, please adjust accordingly.
This is a simple example which covers all cases:

```php
SetNodeReferences::create(
    $node->workspaceName,
    $node->aggregateId,
    $node->originDimensionSpacePoint,
    NodeReferencesToWrite::create(
        // sets the 3 nodes as references
        NodeReferencesForName::fromTargets(
            ReferenceName::fromString('first-reference'),
            NodeAggregateIds::fromArray(['node-aggregate-id-1', 'node-aggregate-id-2', 'node-aggregate-id-3'])
        ),
        // unset previously set references (as we do NOT merge the values)
        NodeReferencesForName::createEmpty(
            ReferenceName::fromString('second-reference')
        ),
        // create a simple node reference and additionally a reference with an additional property
        NodeReferencesForName::fromReferences(
            ReferenceName::fromString('third-reference'),
            [
                NodeReferenceToWrite::fromTarget(NodeAggregateId::fromString('node-aggregate-id-6')),
                NodeReferenceToWrite::fromTargetAndProperties(NodeAggregateId::fromString('node-aggregate-id-5'), PropertyValuesToWrite::fromArray([
                    'propertyOnReference' => 'Hi im living on the edge ;)'
                ])),
            ])
        )
    )
)
```

kitsunet committed:
FEATURE: References on creation and Copy

This reworks references so that multiple reference properties can be
set via a single command and also references can be attached to
`CreateNodeAggregateWithNode` which is also used for copying nodes.

kitsunet commented:
Do not merge yet, probably want to refactor some of the code in the DTOs.  I think we should make the "collection of references for one reference property" an explicit DTO, as the code like this feels a bit too messy for my taste...

kitsunet commented:
On the upside the somewhat weird and unfinished reference snapshots can be removed with this.

kitsunet commented:
And forgot to adapt legacy migration tests (which rely on specific event payloads, see below, before I adjust those we should decide which way to go with b/c)...

kitsunet commented:
oh and I should mention that at the moment this is not forgiving to the existing events. We _could_ do this in the respective fromArray constructors of the DTOs and just "upcast" from the old structure on the fly without much problem, or add a migration for the events.

mhsdesign commented on Neos.ContentRepository.Core/Classes/Feature/Common/NodeReferencingInternals.php:
```diff
@@ -0,0 +1,43 @@
+    private function mapNodeReferencesToSerializedNodeReferences(NodeReferencesToWrite $references, NodeTypeName $nodeTypeName): SerializedNodeReferences
+    {
+        $serializedReferences = [];
+        foreach ($references->getIterator() as $reference) {
```
```suggestion
        foreach ($references as $reference) {
```

isnt it part of the `IteratorAggregate` thing that one doesnt have to do it?

kitsunet replied:
right, right, I messed with this a bit, there are probably more artifacts like this in there.

mhsdesign commented on Neos.ContentRepository.Core/Classes/Feature/NodeReferencing/Dto/NodeReferenceNameToEmpty.php:
```diff
@@ -0,0 +1,30 @@
+ *
+ * @api used as part of commands
+ */
+final readonly class NodeReferenceNameToEmpty
```
hu, i thought we wouldnt need a `*Unset` class like this as only the references will be set that are part of the `SetNodeReferences` command and passing an empty array will unset all references?

Also i had to read twice before i understood `ToEmpty` literally ... im not super happy with it :D

kitsunet replied:
Yep naming TBD, I am not super sold on this either, AND if we introduced the "in between" DTO as said above we wouldn't need this but would just say "NodeReferencesForProperty" and have that be empty for the specific reference name.

kitsunet replied:
this has been removed

mhsdesign commented on Neos.ContentRepository.Core/Classes/Feature/NodeReferencing/Dto/NodeReferenceToWrite.php:
```diff
@@ -26,32 +27,37 @@
+     * @return array{"target": NodeAggregateId, "properties": PropertyValuesToWrite|null}
      */
-    public function jsonSerialize(): array
+    public function targetAndPropertiesToArray(): array
```
i would rather like to inline this code to the caller instead of here?
or a generic `toArray`? Idk or make it internal :D

kitsunet replied:
yeah internal I guess... But htis might again become easier with the middle DTO.

kitsunet replied:
This has been simplified

mhsdesign commented on Neos.ContentRepository.Core/Classes/Feature/NodeReferencing/Dto/NodeReferencesToWrite.php:
```diff
@@ -47,32 +66,34 @@ public static function fromReferences(array $references): self
 
     /**
-     * @param array<int,array<string,mixed>> $values
+     * @param array<string, array<array{"referenceName": string, "target": string, "properties"?: array<string, mixed>}>> $namesAndReferences
```
im not sure if we really like to start annotating the fromArray shape as its a lost battle in terms of time to invest :D But thanks either way ^^

mhsdesign commented on Neos.ContentRepository.Core/Classes/Feature/NodeReferencing/Dto/NodeReferenceToWrite.php:
```diff
@@ -26,32 +27,37 @@
  * @api used as part of commands
  */
-final readonly class NodeReferenceToWrite implements \JsonSerializable
+final readonly class NodeReferenceToWrite
```
thanks for improving the api for setting references ill look into it how this new way is to be used instead and what is to be migrated for Neos 9 users...

While fine tuning the references api id like to also note that it often came to my mind that it didnt seem to have gotten the same love as setting properties. I think both setting references and properties should feel similar and not be totally a different thing.

mhsdesign reviewed (commented):
Thanks for tackling this, some early thoughts before ill play with this ;)

kitsunet commented:
<img width="977" alt="Screenshot 2024-06-18 at 08 47 52" src="https://github.com/neos/neos-development-collection/assets/324408/4ac3678f-66b5-423c-beae-7eed3fc6a666">

I think this would be much nicer code wise (obviously duplicated for Serialized...), bit annoying to use 3 objects for this, but I think it will make the code much better to reason with and empty is then really just the new "middle DTO" with createEmpty()

kitsunet committed:
Fix merge errors and syntax problems

kitsunet committed:
More syntax fixes

kitsunet committed:
Remove forgotten debug statement

kitsunet committed:
Fix new unit Tests

kitsunet committed:
Fix another new unit test

kitsunet committed:
Adjust new behat test with References

kitsunet committed:
Fix broken export test

kitsunet committed:
Merge branch '9.0' into feature/node-reference-copy

# Conflicts:
#	Neos.ContentGraph.DoctrineDbalAdapter/src/DoctrineDbalContentGraphProjection.php

mhsdesign commented:
I remember we discussed this some time ago so i was wondering where it goes from now? I think we discussed for an alternative way than to use a "hacky" NodeReferenceNameToEmpty?

kitsunet committed:
Merge remote-tracking branch 'origin/9.0' into feature/node-reference-copy

kitsunet committed:
Cleaner References write DTOs

kitsunet committed:
Empty properties in refs without special treatment

kitsunet committed:
Merge branch '9.0' into feature/node-reference-copy

kitsunet committed:
Cleanup

kitsunet committed:
Add event migration for references

kitsunet committed:
Merge remote-tracking branch 'origin/9.0' into feature/node-reference-copy

mhsdesign commented on Neos.ContentRepositoryRegistry/Classes/Service/EventMigrationService.php:
```diff
@@ -240,6 +240,107 @@ public function migratePropertiesToUnset(\Closure $outputFn): void
+        $this->copyEventTable($backupEventTableName);
+
+        $streamName = VirtualStreamName::all();
+        $eventStream = $this->eventStore->load($streamName);
```
we should filter here by type `NodeReferencesWereSet` so its cheaper :)

mhsdesign replied:
fyi we cannot use `NodeReferencesWereSet::class` here ... it HAS to be `"NodeReferencesWereSet"` :)

kitsunet committed:
Merge remote-tracking branch 'origin/9.0' into feature/node-reference-copy

mhsdesign commented on Neos.ContentRepository.BehavioralTests/Tests/Behavior/Features/05-NodeReferencing/04-SetNodeReferences_PropertyScopes.feature:
```diff
@@ -63,55 +63,108 @@ Feature: Set node properties with different scopes
+      | specializationsScopedReferences | cs-identifier;source-nodandaise;{"language": "gsw"} | null       |
+
+
+  Scenario: Set node properties in single command
```
```suggestion
  Scenario: Set node references in single command
```

mhsdesign commented on Neos.ContentRepository.BehavioralTests/Tests/Behavior/Features/05-NodeReferencing/04-SetNodeReferences_PropertyScopes.feature:
```diff
@@ -63,55 +63,108 @@ Feature: Set node properties with different scopes
       | targetOrigin    | {"language":"gsw"}  |
 
-  Scenario: Set node properties
+  Scenario: Set node properties in separate commands
```
```suggestion
  Scenario: Set node references in separate commands
```

mhsdesign commented on Neos.ContentRepository.BehavioralTests/Tests/Behavior/Features/NodeCopying/CopyNode_NoDimensions.feature:
```diff
@@ -65,3 +67,24 @@ Feature: Copy nodes (without dimensions)
 
     Then I expect node aggregate identifier "sir-nodeward-nodington-iii-copy" to lead to node cs-identifier;sir-nodeward-nodington-iii-copy;{}
+
+  Scenario: Copy References
```
nice :D and does it also work across dimensions? see https://github.com/neos/neos-development-collection/issues/4614 ...it should work right?

mhsdesign commented on Neos.ContentRepository.Core/Classes/Feature/Common/NodeReferencingInternals.php:
```diff
@@ -0,0 +1,58 @@
+use Neos\ContentRepository\Core\NodeType\NodeType;
+use Neos\ContentRepository\Core\NodeType\NodeTypeName;
+
+/**
+ *
+ */
```
```suggestion
```

mhsdesign commented on Neos.ContentRepository.Core/Classes/Feature/NodeCreation/Command/CreateNodeAggregateWithNode.php:
```diff
@@ -34,7 +35,7 @@
 final readonly class CreateNodeAggregateWithNode implements CommandInterface
 {
-    /**
+    /**x
```
```suggestion
    /**
```

sneaky marker :D

mhsdesign commented on Neos.ContentRepository.Core/Classes/Feature/NodeCreation/Command/CreateNodeAggregateWithNode.php:
```diff
@@ -55,6 +57,7 @@ private function __construct(
         public ?NodeAggregateId $succeedingSiblingNodeAggregateId,
         public ?NodeName $nodeName,
         public NodeAggregateIdsByNodePaths $tetheredDescendantNodeAggregateIds,
+        public ?NodeReferencesToWrite $references,
```
i think this should not be nullable and like `PropertyValuesToWrite` ... that should be easier to handle

mhsdesign commented on Neos.ContentRepository.Core/Classes/Feature/NodeCreation/Command/CreateNodeAggregateWithNodeAndSerializedProperties.php:
```diff
@@ -58,7 +60,8 @@ private function __construct(
         public ?NodeName $nodeName,
-        public NodeAggregateIdsByNodePaths $tetheredDescendantNodeAggregateIds
+        public NodeAggregateIdsByNodePaths $tetheredDescendantNodeAggregateIds,
+        public ?SerializedNodeReferences $references,
```
i think it should not be nullable, but we can just use an empty instance here

mhsdesign commented on Neos.ContentRepository.Core/Classes/Feature/NodeDuplication/Dto/NodeReferenceSnapshot.php:
nice

mhsdesign reviewed (commented):
ill re review later:)

mhsdesign commented on Neos.ContentRepository.Core/Classes/Feature/NodeReferencing/Dto/SerializedNodeReferencesForName.php:
```diff
@@ -0,0 +1,79 @@
+ *
+ * @internal implementation detail of {@see SerializedNodeReferences}
+ */
+class SerializedNodeReferencesForName
```
```suggestion
final readonly class SerializedNodeReferencesForName
```

mhsdesign commented on Neos.ContentRepository.Core/Classes/Feature/NodeReferencing/Dto/SerializedNodeReference.php:
```diff
@@ -20,35 +20,46 @@
  * "Raw" / Serialized node reference as saved in the event log // in projections.
  *
- * @internal
+ * @internal implementation detail of {@see SerializedNodeReferences}
```
```suggestion
 * @api used in commands and events {@see SerializedNodeReferences}
```

mhsdesign commented on Neos.ContentRepositoryRegistry/Classes/Service/EventMigrationService.php:
lecker heute wird wieder array Suppe gekocht :D

mhsdesign commented on Neos.ContentRepository.LegacyNodeMigration/Classes/NodeDataToEventsProcessor.php:
```diff
@@ -474,6 +478,9 @@ private function createNodeVariant(NodeAggregateId $nodeAggregateId, OriginDimen
             );
         }
+
+        // TODO: We should also set references here, shouldn't we?
```
todo

kitsunet replied:
Nod, but I really do not know. I can't fix this and it wasn't the case before so 🤷 not worse than before...

kitsunet committed:
Code cleanup

kitsunet committed:
Code cleanup

mhsdesign committed:
TASK: Fix migrateReferencesToMultiFormat

kitsunet commented on Neos.ContentRepositoryRegistry/Classes/Service/EventMigrationService.php:
```diff
@@ -253,11 +253,11 @@ public function migrateReferencesToMultiFormat(\Closure $outputFn): void
 
         $streamName = VirtualStreamName::all();
-        $eventStream = $this->eventStore->load($streamName, EventStreamFilter::create(EventTypes::create(EventType::fromString(NodeReferencesWereSet::class))));
+        $eventStream = $this->eventStore->load($streamName, EventStreamFilter::create(EventTypes::create(EventType::fromString('NodeReferencesWereSet'))));
```
Somehow this API seems not really nice... But yeah I misread the other place we use it

mhsdesign replied:
haha jup @bwaidelich and me already "complained" about that too :D

bwaidelich replied:
yes, can be improved in the eventstore package to allow for simple types, but that's something for another day

kitsunet commented on Neos.ContentRepositoryRegistry/Classes/Service/EventMigrationService.php:
```diff
@@ -253,11 +253,11 @@ public function migrateReferencesToMultiFormat(\Closure $outputFn): void
             $outputRewriteNotice = fn(string $message) => $outputFn(sprintf('%s@%s %s', $eventEnvelope->sequenceNumber->value, $eventEnvelope->event->type->value, $message));
             if ($eventEnvelope->event->type->value !== 'NodeReferencesWereSet') {
-                continue;
+                throw new \RuntimeException(sprintf('Unhandled event: %s', $eventEnvelope->event->type->value));
```
we could _probably_ even skip the check altogether now with the filter?

mhsdesign replied:
jup but when writing these kind of things i prefer to have this as sanity check :D helps sometimes ^^

mhsdesign committed:
TASK: Simplify code by using automatic `jsonSerialize` like in other places

mhsdesign committed:
TASK: Use variadics to verify array list shape

mhsdesign committed:
TASK: Introduce `fromNameAndTargets` utility and accept flat list in `fromReferences`

mhsdesign committed:
TASK: Remove obsolet serialisation logic from vo's that are not persisted

mhsdesign committed:
TASK: Introduce `NodeReferencesToWrite::withReference` and make merge override previous values

mhsdesign committed:
TASK: Validate `NodeReferencesForName` to not contain duplicate node aggregate ids to write for ONE reference name

mhsdesign committed:
TASK: Simplify `SerializedNodeReferences` capabilities

mhsdesign committed:
TASK: inline `fromReadReferences` so its @internal

bwaidelich commented on Neos.ContentGraph.DoctrineDbalAdapter/src/DoctrineDbalContentGraphProjection.php:
```diff
@@ -597,21 +599,29 @@ function (NodeRecord $node) use ($eventEnvelope) {
+                    $this->dbal->delete($this->tableNames->referenceRelation(), [
+                        'nodeanchorpoint' => $nodeAnchorPoint?->value,
+                        'name' => $referenceName->value
+                    ]);
```
Couldn't this be sth like
```
DELETE FROM $table WHERE name IN (names)
```

bwaidelich commented on Neos.ContentGraph.PostgreSQLAdapter/src/Domain/Projection/Feature/NodeReferencing.php:
```diff
@@ -50,24 +50,25 @@ function (NodeRecord $node) {
-                    $referenceRecord->addToDatabase($this->getDatabaseConnection(), $this->tableNamePrefix);
-                    $position++;
+                foreach ($event->references as $referencesForProperty) {
+                    $this->getDatabaseConnection()->delete($this->tableNamePrefix . '_referencerelation', [
+                        'sourcenodeanchor' => $anchorPoint->value,
+                        'name' => $referencesForProperty->referenceName->value
+                    ]);
+
+                    foreach ($referencesForProperty->references as $reference) {
+                        // set new
+                        $referenceRecord = new ReferenceRelationRecord(
+                            $anchorPoint,
+                            $referencesForProperty->referenceName,
+                            $position,
+                            $reference->properties,
+                            $reference->targetNodeAggregateId
+                        );
+                        $referenceRecord->addToDatabase($this->getDatabaseConnection(), $this->tableNamePrefix);
+                        $position++;
+                    }
```
Same here, can't we turn this into two atomic queries?

mhsdesign replied:
for me the pg adapter is out of scope and there are no tests ... but i fixed it for doctrine

bwaidelich commented on Neos.ContentRepository.Core/Classes/Feature/NodeDuplication/Dto/NodeSubtreeSnapshot.php:
```diff
@@ -105,8 +110,26 @@ public static function fromArray(array $array): self
         );
     }
+
+    private static function serialiseGraphReferences(References $references): SerializedNodeReferences
```
nitpick: in other places we use US english (serialise vs serialize). Furthermore @mhsdesign suggests to rename this:
```suggestion
    private static function serializeProjectedReferences(References $references): SerializedNodeReferences
```

bwaidelich commented on Neos.ContentRepository.Core/Classes/Feature/NodeReferencing/Dto/NodeReferencesForName.php:
```diff
@@ -0,0 +1,62 @@
+use Neos\ContentRepository\Core\SharedModel\Node\ReferenceName;
+
+/**
+ * @api used as part of commands
```
some more docs about what this is about would be nice :)

bwaidelich commented on Neos.ContentRepository.Core/Classes/Feature/NodeReferencing/NodeReferencing.php:
```diff
@@ -112,52 +102,55 @@ private function handleSetSerializedNodeReferences(
+            $this->requireNodeTypeToDeclareReference($sourceNodeAggregate->nodeTypeName, $referenceName);
+            $scopeDeclaration = $sourceNodeType->getReferences()[$referenceName->value]['scope'] ?? '';
+            $scope = PropertyScope::tryFrom($scopeDeclaration) ?: PropertyScope::SCOPE_NODE;
+            // TODO: Optimize affected sets into one event
```
Is there a specific reason why we _don't_?
The current state would be like publishing multiple `NodePropertiesWereSet` events for each property.

Couldn't we just build up the `$referencesByName` array and then build the event outside of the loop?

mhsdesign replied:
Adjusted:)

bwaidelich reviewed (commented):
Great work.. and a lot of it, yikes.
Some rather minor issues

mhsdesign commented:
and the test has to be adjusted as it fails now earlier:

```
001 Scenario: Node reference cannot hold multiple targets to the same node                                          # Features/05-NodeReferencing/01-SetNodeReferences_ConstraintChecks.feature:190
      Then the last command should have thrown an exception of type "InvalidArgumentException" with code 1700150910 # Features/05-NodeReferencing/01-SetNodeReferences_ConstraintChecks.feature:195
        Expected exception code 1700150910, got exception code 1730365958 instead; Message: Duplicate entry in references to write. Target "anthony-destinode" already exists in collection.
        Failed asserting that 1730365958 is identical to 1700150910.
```

mhsdesign committed:
Merge remote-tracking branch 'origin/9.0' into feature/node-reference-copy

mhsdesign committed:
TASK: Adjust documentation and namings

mhsdesign committed:
TASK: Adjust namings to remove obsolete `fromNameAnd` part

and remove duplicated `fromReferences` naming in chain, now its `create` for root level

mhsdesign committed:
TASK: Ensure that `SerializedNodeReferencesForName` are serialized as list

mhsdesign commented:
Okay i fixed the tests again and added some documentation ... i also got rid of the duplicate `NodeReferencesToWrite::fromReferences` and its now just `create` and also the `fromNameAnd` parts are out for the `NodeReferencesForName` as well its guaranteed to stick to a name and thus makes it a little shorter and hopefully easy to use.

```php
SetNodeReferences::create(
    $node->workspaceName,
    $node->aggregateId,
    $node->originDimensionSpacePoint,
    NodeReferencesToWrite::create(
        // sets the 3 nodes as references
        NodeReferencesForName::fromTargets(
            ReferenceName::fromString('first-reference'),
            NodeAggregateIds::fromArray(['node-aggregate-id-1', 'node-aggregate-id-2', 'node-aggregate-id-3'])
        ),
        // unset previously set references (as we do NOT merge the values)
        NodeReferencesForName::createEmpty(
            ReferenceName::fromString('second-reference')
        ),
        // create a simple node reference and additionally a reference with an additional property
        NodeReferencesForName::fromReferences(
            ReferenceName::fromString('third-reference'),
            [
                NodeReferenceToWrite::fromTarget(NodeAggregateId::fromString('node-aggregate-id-6')),
                NodeReferenceToWrite::fromTargetAndProperties(NodeAggregateId::fromString('node-aggregate-id-5'), PropertyValuesToWrite::fromArray([
                    'propertyOnReference' => 'Hi im living on the edge ;)'
                ])),
            ])
    )
)
```

mhsdesign committed:
TASK: Make atomic delete instead of deleting old references one by one

mhsdesign committed:
TASK: Optimise reference writing by splitting by scope

... like in set serialized properties.

That way fewer `NodeReferencesWereSet` are emitted.
At most only one per scope of `PropertyScope`

Also the `PropertyScope` enum was marked internal as its not exposed in apis and no use for the user.

mhsdesign reviewed (approved):
Okay i further made the optimisations basti suggested. Now its truly finished id say <3

Thank you so much for your effort!
